subxt/src/backend/mod.rs

```diff
@@ -392,121 +434,212 @@ mod test {
+                        }
+                    }
+                    Message::Single(item) => self.sender.send(item).await.unwrap(),
+                };
             }
 
-            async fn read(&mut self) -> RpcResult<Vec<Item>> {
-                self.receiver.recv().await.unwrap()
-            }
+            pub async fn write_delayed(&self, items: Message<Item>) {
+                let sender = self.sender.clone();
+                tokio::spawn(async move {
+                    tokio::time::sleep(Duration::from_millis(500)).await;
 
-            async fn write(&self, items: RpcResult<Vec<Item>>) {
-                self.sender.send(items).await.unwrap()
+                    match items {
+                        Message::Many(items) => {
+                            for i in items.unwrap() {
+                                let _ = sender.send(i).await;
+                            }
+                        }
+                        Message::Single(item) => sender.send(item).await.unwrap(),
+                    };
+                });
             }
         }
 
         struct Data {
-            request: MockDataTable,
-            subscription: Subscription,
+            data_table: MockDataTable,
+            subscription_channel: Option<Subscription>,
+            subscription_handlers: HashMap<String, SubscriptionHandler>,
+            method_handlers: HashMap<String, MethodHandler>,
+        }
+
+        impl Data {
+            async fn call_meth<'a>(
+                &'a mut self,
+                meth: &str,
+                params: Option<Box<serde_json::value::RawValue>>,
+            ) -> super::rpc::RawRpcFuture<'a, Box<serde_json::value::RawValue>> {
+                let method = self.method_handlers.get(meth).unwrap_or_else(|| {
+                    panic!("no method named {} registered. Params: {:?}", meth, params)
+                });
+
+                (*method)(&mut self.data_table, &mut self.subscription_channel, params)
+            }
+
+            async fn call_subscription<'a>(
+                &'a mut self,
+                sub: &str,
+                params: Option<Box<serde_json::value::RawValue>>,
+            ) -> super::rpc::RawRpcFuture<'a, super::rpc::RawRpcSubscription> {
+                let sub = self.subscription_handlers.get(sub).unwrap_or_else(|| {
+                    panic!(
+                        "no subscription named {} registered. Params: {:?}",
+                        sub, params
+                    )
+                });
+
+                (*sub)(&mut self.data_table, &mut self.subscription_channel, params)
+            }
+        }
+        pub struct MockRpcBuilder {
+            data: Data,
+        }
+
+        impl MockRpcBuilder {
+            pub fn new() -> Self {
+                Self {
+                    data: Data {
+                        data_table: MockDataTable::new(),
+                        subscription_channel: None,
```

Comment:
is this always `None`? can we remove it?

Reply:
we set it to `Some` when we get a subscription going via "chainHead_v1_follow"